napari/_vispy/layers/base.py

```diff
@@ -122,6 +125,33 @@ def _on_blending_change(self):
         self.node.set_gl_state(**blending_kwargs)
         self.node.update()
 
+    def _on_overlays_change(self):
+        # avoid circular import; TODO: fix?
+        from napari._vispy.utils.visual import create_vispy_overlay
+
+        overlay_models = self.layer._overlays.values()
+
+        for overlay in overlay_models:
+            if overlay not in self.overlays:
+                overlay_visual = create_vispy_overlay(
+                    overlay, layer=self.layer
+                )
+                self.overlays[overlay] = overlay_visual
+                if isinstance(overlay, CanvasOverlay):
+                    overlay_visual.node.parent = (
+                        self.node.parent.parent
+                    )  # viewbox
+                elif isinstance(overlay, SceneOverlay):
+                    overlay_visual.node.parent = self.node
+
+                overlay_visual.node.parent = self.node
+                overlay_visual.reset()
+
+        for overlay in self.overlays.keys():
```

Comment:
```suggestion
        for overlay in self.overlays:
```

Reply:
Actually, reverting this: it was necessary because the dict might change during iteration in the lines below.

Reply:
Then you should list(self.overlays). There is no guarantee that `keys()` will work when mutating dict.

Reply:
TIL! Every day `keys()` becomes more pointless :P

Reply:
As far as I understand, they are useful in the place where you need to pass an iterator. And also backward compatibility.